client/state/stats/chart-tabs/reducer.js

```diff
@@ -19,30 +19,43 @@ import { QUERY_FIELDS } from './constants';
  * @param  {Object} action Action payload
  * @return {Object}        Updated state
  */
-export function counts( state = {}, action ) {
-	switch ( action.type ) {
-		case STATS_CHART_COUNTS_RECEIVE: {
-			const ID = 'period';
-			const records = get( state, `${ action.siteId }.${ action.period }`, [] ).slice();
-			const recordIds = records.map( count => count[ ID ] );
+export const counts = keyedReducer(
+	'siteId',
+	keyedReducer( 'period', ( state = [], action ) => {
+		switch ( action.type ) {
+			case STATS_CHART_COUNTS_RECEIVE: {
+				const ID = 'period';
+				let areThereChanges = false;
+
+				const newState = action.data.reduce(
+					( nextState, recordFromApi ) => {
+						const index = nextState.findIndex( entry => entry[ ID ] === recordFromApi[ ID ] );
+						if ( index >= 0 ) {
+							const newRecord = { ...nextState[ index ], ...recordFromApi };
+							if ( ! isEqual( newRecord, recordFromApi ) ) {
+								areThereChanges = true;
+								nextState[ index ] = newRecord;
+							}
+						} else {
+							areThereChanges = true;
+							nextState.push( recordFromApi );
+						}
+						return nextState;
+					},
+					[ ...state ]
+				);
 
-			// Merge existing records with records from API
-			action.data.forEach( recordFromApi => {
-				const index = recordIds.indexOf( recordFromApi[ ID ] );
-				if ( index >= 0 ) {
-					records[ index ] = { ...records[ index ], ...recordFromApi };
-				} else {
-					records.push( recordFromApi );
+				// Avoid changing state if nothing's changed.
+				if ( ! areThereChanges ) {
+					return state;
 				}
-			} );
 
-			const newState = { ...state };
-			set( newState, `${ action.siteId }.${ action.period }`, records );
-			return newState;
+				return newState;
+			}
 		}
-	}
-	return state;
-}
+		return state;
```

Comment:
I would personally write `return areThereChanges ? newState : state` here 🙂

Reply:
Good point, the whole thing is much simpler now :)